c_emulator/riscv_callbacks_if.h

```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include "sail.h"
+
+// This class is not declared in "riscv_callbacks.h" as that file is
+// included in the C model generated by Sail.
+
+class callbacks_if {
+public:
+  virtual ~callbacks_if() { };
+
+  virtual void mem_write_callback(const char *type, sbits paddr, uint64_t width,
+                                  lbits value)
+      = 0;
+  virtual void mem_read_callback(const char *type, sbits paddr, uint64_t width,
+                                 lbits value)
+      = 0;
+  virtual void mem_exception_callback(sbits paddr, uint64_t num_of_exception)
+      = 0;
+  virtual void xreg_full_write_callback(const_sail_string abi_name,
+                                        unsigned reg, sbits value)
+      = 0;
+  virtual void freg_write_callback(unsigned reg, sbits value) = 0;
+  virtual void csr_full_write_callback(const_sail_string csr_name, unsigned reg,
+                                       sbits value)
+      = 0;
+  virtual void csr_full_read_callback(const_sail_string csr_name, unsigned reg,
+                                      sbits value)
+      = 0;
+  virtual void vreg_write_callback(unsigned reg, lbits value) = 0;
+  virtual void pc_write_callback(sbits value) = 0;
+  virtual void trap_callback() = 0;
+};
+
+void register_callback_if(callbacks_if *cbi);
+void remove_callback_if(callbacks_if *cbi);
```

Comment:
```suggestion
void register_callback(callbacks_if *cbi);
void remove_callback(callbacks_if *cbi);
```

I'm not sure we need the _if on the type either, it's quite obviously an interface.